### PR TITLE
metrics: add metrics for top sql

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs <crazycs520@gmail.com>

### What problem does this PR solve?

As the title said.

### What is changed and how it works?

N/A

### Related changes

- N/A

### Check List 

Tests 

- Unit test
- Integration test
- Manual test
  - check the metrics in Prometheus.

Side effects

- N/A

### Release note 

- No release note.

### CODE REVIEW

File: metrics/metrics.go [hunk 1]
```diff
@@ -153,6 +153,11 @@ func RegisterMetrics() {
 	prometheus.MustRegister(SmallTxnWriteDuration)
 	prometheus.MustRegister(TxnWriteThroughput)
 	prometheus.MustRegister(LoadSysVarCacheCounter)
+	prometheus.MustRegister(RegisterSQLFail)
```
Comment (dragonly) [inline metrics/metrics.go:156]: How about adding a `TopSQL` prefix to all those new variables?

File: metrics/topsql.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package metrics
+
+import "github.com/prometheus/client_golang/prometheus"
+
+// Top SQL metrics.
+var (
+	RegisterSQLFail = prometheus.NewCounter(
```
Comment (dragonly) [inline metrics/topsql.go:7]:
```suggestion
	RegisterSQL = prometheus.NewCounter(
```
How about use a single counter and different labels like `success`/`dropped` to distinguish different results count for the common function `RegisterSQL`?

File: metrics/metrics.go (hunk 1 above)
Comment (crazycs520) [reply metrics/metrics.go]: done.

File: metrics/topsql.go (hunk 2 above)
Comment (crazycs520) [reply metrics/topsql.go]: done.

File: metrics/topsql.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+			Name:      "ignored_total",
+			Help:      "Counter of top-sql ignored register-sql, register-plan, collect-data and report-data, normally it should be 0.",
+		}, []string{LblType})
+
+	TopSQLReportDurationHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "tidb",
+			Subsystem: "topsql",
+			Name:      "report_duration_seconds",
+			Help:      "Bucketed histogram of reporting time (s) of top-sql agent do report",
```
Comment (dragonly) [inline metrics/topsql.go:20]:
```suggestion
			Help:      "Bucket histogram of reporting gRPC duration to the top-sql agent",
```

File: metrics/topsql.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+			Buckets:   prometheus.ExponentialBuckets(0.0005, 2, 29), // 0.5ms ~ 1.5days
+		}, []string{LblResult})
+
+	TopSQLReportDataTotalCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "tidb",
+			Subsystem: "topsql",
+			Name:      "report_data_total",
+			Help:      "Counter of top-sql report records/sql/plan data.",
```
Comment (dragonly) [inline metrics/topsql.go:29]:
```suggestion
			Help:      "Counter of reported records/sql/plan gRPC messages.",
```

File: metrics/topsql.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import "github.com/prometheus/client_golang/prometheus"
+
+// Top SQL metrics.
+var (
+	TopSQLIgnoredCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "tidb",
+			Subsystem: "topsql",
+			Name:      "ignored_total",
+			Help:      "Counter of top-sql ignored register-sql, register-plan, collect-data and report-data, normally it should be 0.",
```
Comment (dragonly) [inline metrics/topsql.go:12]:
```suggestion
			Help:      "Counter of ignored top-sql metrics (register-sql, register-plan, collect-data and report-data), normally it should be 0.",
```

**Review by dragonly — COMMENTED** [review]
rest LGTM

File: util/topsql/reporter/reporter.go [hunk 6]
```diff
@@ -133,6 +134,18 @@ func NewRemoteTopSQLReporter(client ReportClient) *RemoteTopSQLReporter {
 	return tsr
 }
 
+var (
+	ignoreRegisterSQLCnt          = metrics.TopSQLIgnoredCounter.WithLabelValues("register-sql")
+	ignoreRegisterPlanCnt         = metrics.TopSQLIgnoredCounter.WithLabelValues("register-plan")
+	ignoreCollectCPUTimeCnt       = metrics.TopSQLIgnoredCounter.WithLabelValues("collect-data")
+	ignoreReportDataCnt           = metrics.TopSQLIgnoredCounter.WithLabelValues("report-data")
+	reportDurationSuccHistogram   = metrics.TopSQLReportDurationHistogram.WithLabelValues("OK")
+	reportDurationFailedHistogram = metrics.TopSQLReportDurationHistogram.WithLabelValues("Error")
```
Comment (breeswish) [inline util/topsql/reporter/reporter.go:143]: If I remember correctly there are constant labels for OK and Error, maybe we can reuse them?

File: util/topsql/reporter/reporter.go [hunk 7]
```diff
@@ -133,6 +134,18 @@ func NewRemoteTopSQLReporter(client ReportClient) *RemoteTopSQLReporter {
...
+	topSQLReportRecordCnt         = metrics.TopSQLReportDataTotalCounter.WithLabelValues("record")
+	topSQLReportSQLCnt            = metrics.TopSQLReportDataTotalCounter.WithLabelValues("sql")
```
Comment (breeswish) [inline util/topsql/reporter/reporter.go:145]: should it be "statement" instead of "sql"?

File: util/topsql/reporter/reporter.go [hunk 8]
```diff
@@ -133,6 +134,18 @@ func NewRemoteTopSQLReporter(client ReportClient) *RemoteTopSQLReporter {
 	return tsr
 }
 
+var (
+	ignoreRegisterSQLCnt          = metrics.TopSQLIgnoredCounter.WithLabelValues("register-sql")
+	ignoreRegisterPlanCnt         = metrics.TopSQLIgnoredCounter.WithLabelValues("register-plan")
+	ignoreCollectCPUTimeCnt       = metrics.TopSQLIgnoredCounter.WithLabelValues("collect-data")
+	ignoreReportDataCnt           = metrics.TopSQLIgnoredCounter.WithLabelValues("report-data")
```
Comment (breeswish) [inline util/topsql/reporter/reporter.go:141]: I would prefer to use `register_sql` to match other labels:

image

File: metrics/topsql.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+			Help:      "Counter of ignored top-sql metrics (register-sql, register-plan, collect-data and report-data), normally it should be 0.",
+		}, []string{LblType})
+
+	TopSQLReportDurationHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "tidb",
+			Subsystem: "topsql",
+			Name:      "report_duration_seconds",
+			Help:      "Bucket histogram of reporting time (s) to the top-sql agent",
+			Buckets:   prometheus.ExponentialBuckets(0.0005, 2, 29), // 0.5ms ~ 1.5days
```
Comment (breeswish) [inline metrics/topsql.go:21]: Is it really necessary to have such long time (1.5 days) recording?

File: metrics/topsql.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+		}, []string{LblResult})
+
+	TopSQLReportDataTotalCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "tidb",
+			Subsystem: "topsql",
+			Name:      "report_data_total",
+			Help:      "Counter of top-sql reported records/sql/plan.",
```
Comment (breeswish) [inline metrics/topsql.go:29]: I think you should also track with how much data each report takes (which is a histogram). And when using a histogram then counter is no longer needed (as a histogram contains a sum counter).

Comment (crazycs520) [reply metrics/topsql.go:29]: done.

File: metrics/topsql.go (hunk 9 above)
Comment (crazycs520) [reply metrics/topsql.go]: nice catch.

File: util/topsql/reporter/reporter.go [hunk 11]
```diff
@@ -155,6 +169,7 @@ func (tsr *RemoteTopSQLReporter) RegisterSQL(sqlDigest []byte, normalizedSQL str
 // This function is thread-safe and efficient.
 func (tsr *RemoteTopSQLReporter) RegisterPlan(planDigest []byte, normalizedBinaryPlan string) {
 	if tsr.planMapLength.Load() >= variable.TopSQLVariable.MaxCollect.Load() {
+		ignoreRegisterPlanCnt.Inc()
```
Comment (breeswish) [inline util/topsql/reporter/reporter.go:172]: maybe better to distinguish it as "ignore_exceed".

File: util/topsql/reporter/reporter.go [hunk 12]
```diff
@@ -178,6 +193,7 @@ func (tsr *RemoteTopSQLReporter) Collect(timestamp uint64, records []tracecpu.SQ
 	}:
 	default:
 		// ignore if chan blocked
+		ignoreCollectCPUTimeCnt.Inc()
```
Comment (breeswish) [inline util/topsql/reporter/reporter.go:196]: ignore_channel_full

File: util/topsql/reporter/client.go [hunk 13]
```diff
@@ -145,6 +148,7 @@ func (r *GRPCReportClient) sendBatchSQLMeta(ctx context.Context, sqlMap *sync.Ma
 	if err != nil {
 		return err
 	}
+	topSQLReportStatementHistogram.Observe(float64(cnt))
 	_, err = stream.CloseAndRecv()
```
Comment (breeswish) [inline util/topsql/reporter/client.go:152]: Should we also observe how long this operation takes?

**Review by breeswish — APPROVED** [review]
Mostly LGTM, optional improvement suggestions.

File: util/topsql/reporter/reporter.go (hunk 11 above)
Comment (crazycs520) [reply util/topsql/reporter/reporter.go]: great, done.

File: util/topsql/reporter/reporter.go (hunk 12 above)
Comment (crazycs520) [reply util/topsql/reporter/reporter.go]: great, done.

File: util/topsql/reporter/client.go (hunk 13 above)
Comment (crazycs520) [reply util/topsql/reporter/client.go:159]: great, done.

File: util/topsql/reporter/reporter.go [hunk 14]
```diff
@@ -133,6 +134,18 @@ func NewRemoteTopSQLReporter(client ReportClient) *RemoteTopSQLReporter {
...
 
+var (
+	ignoreRegisterSQLCnt           = metrics.TopSQLIgnoredCounter.WithLabelValues("register_sql")
+	ignoreRegisterPlanCnt          = metrics.TopSQLIgnoredCounter.WithLabelValues("register_plan")
+	ignoreCollectCPUTimeCnt        = metrics.TopSQLIgnoredCounter.WithLabelValues("collect_data")
+	ignoreReportDataCnt            = metrics.TopSQLIgnoredCounter.WithLabelValues("report_data")
+	reportDurationSuccHistogram    = metrics.TopSQLReportDurationHistogram.WithLabelValues(metrics.LblOK)
+	reportDurationFailedHistogram  = metrics.TopSQLReportDurationHistogram.WithLabelValues(metrics.LblError)
+	topSQLReportRecordHistogram    = metrics.TopSQLReportDataHistogram.WithLabelValues("record")
+	topSQLReportStatementHistogram = metrics.TopSQLReportDataHistogram.WithLabelValues("statement")
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply util/topsql/reporter/reporter.go]: ok

**Review by dragonly — APPROVED** [review]

**Pull request merged**